overlord/devicestate/users.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	// then the UserState
+	st.Lock()
+	u, err := auth.RemoveUserByUsername(st, username)
+	st.Unlock()
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+	return u, nil
+}
+
+// CreateKnownUsers create known user(s)
+// user details are fetched from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If email is passed, only corresponding system user assertion is used for user creation.
+func CreateKnownUsers(st *state.State, mgr *DeviceManager, sudoer bool, email string) ([]CreatedUser, *UserError) {
+	var err error
+	st.Lock()
+	model, err := mgr.Model()
+	st.Unlock()
+	if err != nil {
+		return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get model assertion: %v", err)}
+	}
+	st.Lock()
+	serial, err := mgr.Serial()
+	st.Unlock()
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get serial: %v", err)}
+	}
+
+	st.Lock()
+	db := assertstate.DB(st)
+	st.Unlock()
+
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	if email == "" {
+		return u.createAllKnownSystemUsers(st)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+
+	if err != nil {
+		return nil, &UserError{Internal: false, Err: err}
+	}
+	opts.Sudoer = sudoer
+	createdUser, userErr := u.createUser(st, opts, username, email)
+	var createdUsers []CreatedUser
+	createdUsers = append(createdUsers, createdUser)
+	return createdUsers, userErr
+}
+
+// CreateUser creates linux user based on the passed email
+// username and public ssh keys for the created user account are determined
+// from Ubuntu store based on the passed email.
+func CreateUser(st *state.State, sudoer bool, email string) (CreatedUser, *UserError) {
+	if email == "" {
+		return CreatedUser{}, &UserError{Internal: false, Err: fmt.Errorf("cannot create user: 'email' field is empty")}
+	}
+
+	st.Lock()
+	db := assertstate.DB(st)
+	st.Unlock()
+
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  nil,
+		serialAs: nil,
+		isSudoer: sudoer,
+	}
+
+	st.Lock()
+	storeService := snapstate.Store(st, nil)
+	st.Unlock()
+	username, opts, err := getUserDetailsFromStore(storeService, email)
+
+	if err != nil {
+		return CreatedUser{}, &UserError{Internal: false, Err: err}
+	}
+	opts.Sudoer = sudoer
+	return u.createUser(st, opts, username, email)
+}
+
+func (u *createUserOpts) createUser(st *state.State, opts *osutil.AddUserOptions, username, email string) (CreatedUser, *UserError) {
```

Comment:
it's unusual for this to appear before the struct definition, probably some reordering is needed

Reply:
Reordered the file